ebpf/chanmon.go

```diff
@@ -166,6 +171,69 @@ func processChansendEvents(objs *bpfObjects) error {
 	return nil
 }
 
+func processChanrecvEvents(objs *bpfObjects) error {
+	var key bpfChanrecvEventKey
+	var event bpfChanrecvEvent
+	var keysToDelete []bpfChanrecvEventKey
+	stackIdSetToDelete := make(map[int32]struct{})
+
+	events := objs.ChanrecvEvents.Iterate()
+	for events.Next(&key, &event) {
+		stack, err := extractStack(objs, event.StackId)
+		if err != nil {
+			slog.Warn(err.Error())
+			continue
+		}
+		if _, ok := stackIdSetToDelete[event.StackId]; !ok {
+			stackIdSetToDelete[event.StackId] = struct{}{}
+		}
+		keysToDelete = append(keysToDelete, key)
+		attrs := []any{
+			slog.Int64("goroutine_id", int64(key.GoroutineId)),
+			slog.Int64("stack_id", int64(event.StackId)),
+			slog.Any("stack", stack),
+		}
+		// As of Go version 1.21.3, there is no mechanism to access `selected` and `received`,
+		// which are the first and second return values of `runtime.chanrecv`, respectively.
+		// They are stored in the rax and rbx registers of the amd64 architecture.
+		// Upon examining the assembly code of `runtime.chanrecv` through gdb,
+		// it's evident that the return values aren't stored at the stack pointer,
+		// thereby rendering it impossible to retrieve `selected` and `received` both.
+		// See https://github.com/keisku/chanmon/pull/2
```

Comment:
```shell
sudo gdb /path/to/gobinary
# Opening gdb...
(gdb) b runtime.chanrecv
Breakpoint 1 at 0x405460: file /snap/go/10389/src/runtime/chan.go, line 457.
(gdb) run
(gdb) disas/m
Dump of assembler code for function runtime.chanrecv:
...
466			if !block {
   0x0000000000405594 <+308>:	jne    0x4055a1 <runtime.chanrecv+321>

467				return
   0x0000000000405596 <+310>:	xor    %eax,%eax
   0x0000000000405598 <+312>:	mov    %eax,%ebx
   0x000000000040559a <+314>:	add    $0x68,%rsp
   0x000000000040559e <+318>:	pop    %rbp
   0x000000000040559f <+319>:	nop
   0x00000000004055a0 <+320>:	ret
...
484			if atomic.Load(&c.closed) == 0 {
   0x00000000004054a4 <+68>:	mov    0x1c(%rax),%edx
   0x00000000004054a7 <+71>:	test   %edx,%edx
   0x00000000004054a9 <+73>:	je     0x4054e6 <runtime.chanrecv+134>

485				// Because a channel cannot be reopened, the later observation of the channel
486				// being not closed implies that it was also not closed at the moment of the
487				// first observation. We behave as if we observed the channel at that moment
488				// and report that the receive cannot proceed.
489				return
   0x00000000004054e6 <+134>:	xor    %eax,%eax
   0x00000000004054e8 <+136>:	mov    %eax,%ebx
   0x00000000004054ea <+138>:	add    $0x68,%rsp
   0x00000000004054ee <+142>:	pop    %rbp
   0x00000000004054ef <+143>:	ret
...
502				return true, false
   0x00000000004054d9 <+121>:	mov    $0x1,%eax
   0x00000000004054de <+126>:	xor    %ebx,%ebx
   0x00000000004054e0 <+128>:	add    $0x68,%rsp
   0x00000000004054e4 <+132>:	pop    %rbp
   0x00000000004054e5 <+133>:	ret
...
522				return true, false
   0x0000000000405585 <+293>:	mov    $0x1,%eax
   0x000000000040558a <+298>:	xor    %ebx,%ebx
   0x000000000040558c <+300>:	add    $0x68,%rsp
   0x0000000000405590 <+304>:	pop    %rbp
   0x0000000000405591 <+305>:	ret
...
533				return true, true
   0x0000000000405a11 <+1457>:	mov    $0x1,%eax
   0x0000000000405a16 <+1462>:	mov    %eax,%ebx
   0x0000000000405a18 <+1464>:	add    $0x68,%rsp
   0x0000000000405a1c <+1468>:	pop    %rbp
   0x0000000000405a1d <+1469>:	ret
...
553			return true, true
   0x000000000040564b <+491>:	mov    $0x1,%eax
   0x0000000000405650 <+496>:	mov    %eax,%ebx
   0x0000000000405652 <+498>:	add    $0x68,%rsp
   0x0000000000405656 <+502>:	pop    %rbp
   0x0000000000405657 <+503>:	ret

554		}
555	
556		if !block {
   0x0000000000405662 <+514>:	jne    0x40567a <runtime.chanrecv+538>

557			unlock(&c.lock)
   0x0000000000405664 <+516>:	nop

558			return false, false
   0x0000000000405670 <+528>:	xor    %eax,%eax
   0x0000000000405672 <+530>:	mov    %eax,%ebx
   0x0000000000405674 <+532>:	add    $0x68,%rsp
   0x0000000000405678 <+536>:	pop    %rbp
   0x0000000000405679 <+537>:	ret

559		}
```

Reply:
There is no macro to retrieve value from `%eabx`. 
https://elixir.bootlin.com/linux/v6.2.16/source/tools/lib/bpf/bpf_tracing.h#L73